src/utils/abortableAtom.ts

```diff
@@ -0,0 +1,45 @@
+import {
+  atom,
+  SECRET_INTERNAL_registerPromiseAbort as registerPromiseAbort,
+} from 'jotai'
+import type { Atom, WritableAtom } from 'jotai'
+
+type Getter = Parameters<Atom<unknown>['read']>[0]
+type WriteGetter = Parameters<WritableAtom<unknown, unknown>['write']>[0]
+type Setter = Parameters<WritableAtom<unknown, unknown>['write']>[1]
+
+type Read<Value> = (get: Getter, options: { signal: AbortSignal }) => Value
+
+type Write<Update, Result extends void | Promise<void>> = (
+  get: WriteGetter,
+  set: Setter,
+  update: Update
+) => Result
+
+// writable derived atom
+export function abortableAtom<
+  Value,
+  Update,
+  Result extends void | Promise<void> = void
+>(
+  read: Read<Value>,
+  write: Write<Update, Result>
+): WritableAtom<Value, Update, Result>
+
+// read-only derived atom
+export function abortableAtom<Value>(read: Read<Value>): Atom<Value>
+
+export function abortableAtom<
+  Value,
+  Update,
+  Result extends void | Promise<void>
+>(read: Read<Value>, write?: Write<Update, Result>) {
+  const { signal, abort } = new AbortController()
```

Comment:
I am not a zustand user, so may be my comment will be irrelevant 🙏


AbortController only works with fetch and axios and may be some other libraries.

Cancellations in general affect any asynchronous behavior: browser api (timeout, interval), Promises, async/await .. I suggest supporting a mechanism that allows cancellations of all kinds, and not instantiating an abort controller at each run.

In a library I am preparing, I solved this by allowing the user's function to register any amount of abort callbacks, that are invoked either by user action or by effect cleanup.


```
function myStateProducer(props) {
  props.onAbort(() => clearTimeout(id));
  props.onAbort(() => clearInterval(id));
  props.onAbort((reason) => controller.abort(reason));
}
```

This allows the user to easily implement cancellations of all kinds rather that relying on AbortController  which isn't supported in legacy browsers.

How can we cancel atoms ? is there any `abort` function provided ?

Reply:
As far as I understand, AbortController is a general subscriber, and has nothing to do with fetch and others. We can use it for anything. setTimeout/clearTimeout can work too. (DX-wise, not sure if it's ideal) It would be nice to have such usage in docs.

Reply:
My plan was to use `signal.aborted`. It's not super nice, though. Maybe there would be better solutions. What we keep in mind is we don't often directly use promises. In most cases we just use `async` functions.

I'm not sure if we are on the same page. Feel free to try your own idea. It can be AC based, or independent. We shouldn't need to change jotai core in either case.

(I feel like I'm missing some cases like complex dependency graph...